### PR TITLE
wolfSSL Compatibility support for OpenVPN

### PR DESCRIPTION
**Pull request by dgarske**
wolfSSL Compatibility support for OpenVPN

* Adds compatibility API's for:
	* `sk_ASN1_OBJECT_free`
	* `sk_ASN1_OBJECT_num`
	* `sk_ASN1_OBJECT_value`
	* `sk_X509_OBJECT_num`
	* `sk_X509_OBJECT_value`
	* `sk_X509_OBJECT_delete`
	* `sk_X509_NAME_find`
	* `sk_X509_INFO_free`
	* `BIO_get_len`
	* `BIO_set_ssl`
	* `BIO_should_retry` (stub)
	* `X509_OBJECT_free`
	* `X509_NAME_get_index_by_OBJ`
	* `X509_INFO_free`
	* `X509_STORE_get0_objects`
	* `X509_check_purpose` (stub)
	* `PEM_read_bio_X509_CRL`
	* `PEM_X509_INFO_read_bio`
	* `ASN1_BIT_STRING_new`
	* `ASN1_BIT_STRING_free`
	* `ASN1_BIT_STRING_get_bit`
	* `ASN1_BIT_STRING_set_bit`
	* `DES_check_key_parity`
	* `EC_GROUP_order_bits`
	* `EC_get_builtin_curves`
	* `EVP_CIPHER_CTX_cipher`
	* `EVP_PKEY_get0_EC_KEY`
	* `EVP_PKEY_get0_RSA`
	* `EVP_PKEY_get0_DSA` (stub)
	* `HMAC_CTX_new`
	* `HMAC_CTX_free`
	* `HMAC_CTX_reset`
	* `HMAC_size`
	* `OBJ_txt2obj`
	* `RSA_meth_new`
	* `RSA_meth_free`
	* `RSA_meth_set_pub_enc`
	* `RSA_meth_set_pub_dec`
	* `RSA_meth_set_priv_enc`
	* `RSA_meth_set_priv_dec`
	* `RSA_meth_set_init`
	* `RSA_meth_set_finish`
	* `RSA_meth_set0_app_data`
	* `RSA_get_method_data`
	* `RSA_set_method`
	* `RSA_get0_key`
	* `RSA_set0_key`
	* `RSA_flags`
	* `RSA_set_flags`
	* `RSA_bits`
	* `SSL_CTX_set_ciphersuites`
	* `SSL_CTX_set_security_level` (stub)
	* `SSL_export_keying_material` (stub)
	* `DSA_bits` (stub)
* Changes to support password callback trial and NO_PASSWORD. Replaces PR #2505.
* Renamed `wolfSSL_SSL_CTX_get_client_CA_list` to `wolfSSL_CTX_get_client_CA_list`.
* Cleanup of "sk" compatibility.

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -31185,6 +31421,34 @@ int wolfSSL_EC_GROUP_get_order(const WOLFSSL_EC_GROUP *group,
...
+    }
+
+    ret = mp_init(&order);
+    if (ret == 0) {
+        ret = mp_read_radix(&order, ecc_sets[group->curve_idx].order,
+            MP_RADIX_HEX);
+        if (ret == 0)
+            ret = mp_unsigned_bin_size(&order);
+        if (ret >= 0)
+            ret = mp_to_unsigned_bin(&order, (byte*)&order_bits);
```
Comment (toddouska) [inline src/ssl.c]: This is big endian unsigned format right?  Is that what the caller is expecting?  Did you look at mp_count_bits()?

File: src/ssl.c [hunk 2]
```diff
@@ -7662,6 +7718,32 @@ void wolfSSL_sk_X509_EXTENSION_free(WOLFSSL_STACK* sk)
     XFREE(sk, NULL, DYNAMIC_TYPE_X509);
 }
 
+int wolfSSL_ASN1_BIT_STRING_set_bit(WOLFSSL_ASN1_BIT_STRING* str, int pos,
+    int val)
+{
+    int bytes_cnt = pos/8;
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: check for negative value input for "pos"

File: src/ssl.c [hunk 3]
```diff
@@ -19121,14 +19272,27 @@ int wolfSSL_ASN1_STRING_to_UTF8(unsigned char **out, WOLFSSL_ASN1_STRING *in)
...
-    WOLFSSL_STUB("ASN1_STRING_to_UTF8");
-    return -1;
+    unsigned char* buf;
+    int inLen;
+
+    if (!out || !in) {
+        return -1;
+    }
+
+    inLen = wolfSSL_ASN1_STRING_length(in);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: right now wolfSSL_ASN1_STRING_length always returns 0 or greater. I think it is slightly dangerous to always assume the int returned in the future will be though.

File: src/ssl.c [hunk 4]
```diff
@@ -19121,14 +19272,27 @@ int wolfSSL_ASN1_STRING_to_UTF8(unsigned char **out, WOLFSSL_ASN1_STRING *in)
...
+    buf = (unsigned char*)XMALLOC(inLen + 1, NULL,
+        DYNAMIC_TYPE_OPENSSL);
+    if (!buf) {
+        return -1;
+    }
+    XMEMCPY(buf, wolfSSL_ASN1_STRING_data(in), inLen + 1);
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: wolfSSL_ASN1_STRING_data can return NULL

File: src/ssl.c [hunk 5]
```diff
@@ -32958,6 +33240,102 @@ int wolfSSL_RSA_LoadDer_ex(WOLFSSL_RSA* rsa, const unsigned char* derBuf,
 
     return WOLFSSL_SUCCESS;
 }
+
+WOLFSSL_RSA_METHOD *wolfSSL_RSA_meth_new(const char *name, int flags)
+{
+    int name_len;
+    WOLFSSL_RSA_METHOD* meth;
+
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: null check on "name"

File: src/ssl.c [hunk 6]
```diff
@@ -32958,6 +33240,102 @@ int wolfSSL_RSA_LoadDer_ex(WOLFSSL_RSA* rsa, const unsigned char* derBuf,
...
+        XFREE(meth->name, NULL, DYNAMIC_TYPE_OPENSSL);
+        XFREE(meth, NULL, DYNAMIC_TYPE_OPENSSL);
+    }
+}
+
+int wolfSSL_RSA_meth_set(WOLFSSL_RSA_METHOD *rsa, void* p)
+{
+    (void)rsa;
+    (void)p;
+    WOLFSSL_MSG("RSA_METHOD is not implemented.");
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: please use WOLFSSL_STUB if this is a stub function, and surround with NO_WOLFSSL_STUB macro

File: src/ssl.c [hunk 7]
```diff
@@ -32958,6 +33240,102 @@ int wolfSSL_RSA_LoadDer_ex(WOLFSSL_RSA* rsa, const unsigned char* derBuf,
...
+    return 1;
+}
+
+#if defined(OPENSSL_EXTRA)
+int wolfSSL_RSA_set_method(WOLFSSL_RSA *rsa, WOLFSSL_RSA_METHOD *meth)
+{
+    rsa->meth = meth;
```
Comment (JacobBarthelmeh) [inline src/ssl.c]: null check on "rsa"

**Comment by julek-wolfssl** [comment]
Maybe there should be a configure option like `--enable-openvpn` added as the current list is quite long:
```
./configure --enable-opensslall --enable-keygen --enable-rsapss --enable-des3 \
--enable-dtls --enable-certgen --enable-certreq --enable-certext \
--enable-sessioncerts --enable-crl --enable-ocsp  --enable-aesctr \
CFLAGS="-DWOLFSSL_DES_ECB -DWOLFSSL_LOG_PRINTF -DWOLFSSL_PUBLIC_MP -DHAVE_EX_DATA"
```

**Comment by dgarske** [comment]
> Maybe there should be a configure option like `--enable-openvpn` added as the current list is quite long:
> 
> ```
> ./configure --enable-opensslall --enable-keygen --enable-rsapss --enable-des3 \
> --enable-dtls --enable-certgen --enable-certreq --enable-certext \
> --enable-sessioncerts --enable-crl --enable-ocsp  --enable-aesctr \
> CFLAGS="-DWOLFSSL_DES_ECB -DWOLFSSL_LOG_PRINTF -DWOLFSSL_PUBLIC_MP -DHAVE_EX_DATA"
> ```

Thanks, good suggestion. I'll see about narrowing down the list to just the ones OpenVPN requires and adding an enable option. Some of these are for the other projects such as OpenVSwitch, StrongSwan, WebRTC and IPOP-VPN also being ported.

**Comment by JacobBarthelmeh** [comment]
Thanks David and Juliusz

**Pull request merged**